numpy/core/src/multiarray/multiarraymodule.c

```diff
@@ -1627,12 +1627,31 @@ _array_fromobject_generic(
                 goto finish;
             }
         }
-        /* One more chance */
+        /* One more chance for faster exit if user specified the dtype. */
         oldtype = PyArray_DESCR(oparr);
         if (PyArray_EquivTypes(oldtype, type)) {
             if (copy != NPY_COPY_ALWAYS && STRIDING_OK(oparr, order)) {
                 Py_INCREF(op);
-                ret = oparr;
+                if (oldtype == type) {
+                    ret = oparr;
+                }
+                else {
+                    /* Create a new PyArrayObject from the caller's
+                     * PyArray_Descr. Use the reference `op` as the base
+                     * object. */
+                    Py_INCREF(type);
+                    ret = (PyArrayObject *)PyArray_NewFromDescrAndBase(
+                            Py_TYPE(op),
+                            type,
+                            PyArray_NDIM(oparr),
+                            PyArray_DIMS(oparr),
+                            PyArray_STRIDES(oparr),
+                            PyArray_DATA(oparr),
+                            PyArray_FLAGS(oparr),
+                            op,
+                            op
+                            );
```

Comment:
Oh, either I just didn't think of it, or expected that this steals a reference to `op` for the `base`.  We are missing a DECREF, I will make a PR.

Reply:
Yes, I think I originally tried `PyArray_NewFromDescr` with `op` for `*data` and `nullptr` for `*obj`, and `PyArray_SetBaseObject` or something that steals the reference for the base. Anyway, I guess we were sloppy when switching to `PyArray_NewFromDescrAndBase`. Thanks for the quick catch!